src/NServiceBus.TransportTests/When_an_operation_is_cancelled_during_recoverability.cs

```diff
@@ -0,0 +1,42 @@
+﻿namespace NServiceBus.TransportTests
+{
+    using System;
+    using System.Threading.Tasks;
+    using NUnit.Framework;
+
+    public class When_an_operation_is_cancelled_during_recoverability : NServiceBusTransportTest
+    {
+        [TestCase(TransportTransactionMode.None)]
+        [TestCase(TransportTransactionMode.ReceiveOnly)]
+        [TestCase(TransportTransactionMode.SendsAtomicWithReceive)]
+        [TestCase(TransportTransactionMode.TransactionScope)]
+        public async Task Should_invoke_critical_error(TransportTransactionMode transactionMode)
+        {
+            var recoverabilityStarted = new TaskCompletionSource<bool>();
+            var criticalErrorInvoked = false;
+
+            OnTestTimeout(() => recoverabilityStarted.SetCanceled());
+
+            await StartPump(
+                (_, __) => throw new Exception(),
+                (_, cancellationToken) =>
+                {
+                    recoverabilityStarted.SetResult(true);
+
+                    throw new OperationCanceledException();
+                },
+                transactionMode,
+                (_, __, ___) => criticalErrorInvoked = true);
+
+            await SendMessage(InputQueueName);
+
+            _ = await recoverabilityStarted.Task;
+
+            await Task.Delay(TimeSpan.FromSeconds(1));
```

Comment:
Why a delay? Assuming a race can we use a `TaskCompletionSource` to sync instead?

Reply:
See https://github.com/Particular/NServiceBus/pull/5931#discussion_r577951142